orttraining/orttraining/training_ops/cuda/communication/recv.cc

```diff
@@ -3,75 +3,112 @@
-  cudaLaunchHostFunc(commStream, HostRecv, &info_shape_size);
-  cudaStreamSynchronize(commStream);
-  
-  // Receive shape.
-  std::vector<int64_t> tensor_shape(tensor_shape_size);
-  CommInfo_t info_shape{tensor_shape.data(), tensor_shape.size() * sizeof(int64_t), static_cast<int>(src_), static_cast<int>(tag_)};
-  cudaLaunchHostFunc(commStream, HostRecv, &info_shape);
+  // Receive shape sizes and aggregated size
+  CommInfo_t info_shape_sizes{prefix_tensor_shape_sizes.data(),
+                              tensor_num * static_cast<int>(sizeof(size_t)),
+                              src,
+                              static_cast<int>(tag_)};
+  CommInfo_t info_aggregated_size{&aggregated_aligned_tensor_bytes,
+                                  static_cast<int>(sizeof(size_t)),
+                                  src,
+                                  static_cast<int>(tag_)};
+  cudaLaunchHostFunc(commStream, HostRecv, &info_shape_sizes);
+  cudaLaunchHostFunc(commStream, HostRecv, &info_aggregated_size);
   cudaStreamSynchronize(commStream);
 
-  Tensor* x_tensor = ctx->Output(1, tensor_shape);
-  T* x = x_tensor->MutableData<T>();
-
-  // Receive actual data.
-  IAllocatorUniquePtr<T> buffer = AllocateBufferOnCPUPinned<T>(x_tensor->Shape().Size());
-  CommInfo_t info_data{buffer.get(), x_tensor->Shape().Size() * sizeof(T), static_cast<int>(src_), static_cast<int>(tag_)};
+  // Receive shapes and data buffer
+  aggregated_tensor_shapes.resize(prefix_tensor_shape_sizes[tensor_num - 1]);
+  IAllocatorUniquePtr<char> buffer =
+      AllocateBufferOnCPUPinned<char>(static_cast<size_t>(aggregated_aligned_tensor_bytes));
+  CommInfo_t info_shapes{aggregated_tensor_shapes.data(),
+                         static_cast<int>(aggregated_tensor_shapes.size()) * static_cast<int>(sizeof(int64_t)),
+                         src,
+                         static_cast<int>(tag_)};
+  CommInfo_t info_data{buffer.get(),
+                       static_cast<int>(aggregated_aligned_tensor_bytes),
+                       src,
+                       static_cast<int>(tag_)};
+  cudaLaunchHostFunc(commStream, HostRecv, &info_shapes);
   cudaLaunchHostFunc(commStream, HostRecv, &info_data);
   cudaStreamSynchronize(commStream);
   cudaStreamDestroy(commStream);
 
-  // TODO: we need to avoid copying data to host after enabling CUDA-aware MPI in the build.
-  cudaMemcpy(x, buffer.get(), sizeof(T) * x_tensor->Shape().Size(), cudaMemcpyHostToDevice);
+  // Create Tensors
+  size_t begin = 0;
+  size_t tensor_offset_in_bytes = 0;
+  for (int i = 0; i < tensor_num; ++i) {
+    std::vector<int64_t> tensor_shape(aggregated_tensor_shapes.begin() + begin,
+                                      aggregated_tensor_shapes.begin() + prefix_tensor_shape_sizes[i]);
+    begin = prefix_tensor_shape_sizes[i];
+
+    Tensor* x_tensor = ctx->Output(i + 1, tensor_shape);
+    // Find the next aligned offset in the tensor buffer to meet alignment requirement
```

Comment:
This `finding the next alignment address` is tightly coupled with that one in Send. Could you make it a function and use it in both of Send and Recv?

Reply:
You mean putting them in common?
Sure, why not.